### PR TITLE
[Feature-13511] Submit Spark task directly on Kubernetes

### LINKED ISSUE #13511: [Feature][Task] Submit Spark task directly on Kubernetes
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In [the Spark doc](https://spark.apache.org/docs/latest/running-on-kubernetes.html), now we can directly submit jobs to Kubernetes and schedule the jobs in Pods, so users can directly use the Kubernetes and don't need to deploy yarn anymore. However DolphinScheduler currently hardcoded the master to yarn (`--master yarn`)

https://github.com/apache/dolphinscheduler/blob/0ee82c5492167f01e4ee5f1d55f2f248181c65fe/dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/main/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTask.java#L131-L139

In order to submit jobs to Kubernetes, the `--master` option must be configurable so users can specify it as the Kubernetes API server address.

### Use case

There are at least 2 use cases:

- DolphinScheduler is not deployed inside Kubernetes, but users want to submit jobs to Kubernetes cluster.
- DolphinScheduler is also deployed inside Kubernetes, and users also want to submit jobs to Kubernetes cluster.

### Related issues

https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue+is%3Aopen+spark+kubernetes

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Radeity**
## Purpose of the pull request

- This PR closes https://github.com/apache/dolphinscheduler/issues/13511.

## Brief change log

Now in DS, use `Yarn` as application manager by default, e.g. directly executes `yarn application -kill <appId>` for canceling application. For using different resource manager such as `Kubernetes`, this pr makes following changes:

- Create `ApplicationManager`, implement `KubernetesApplicationManager` and `YarnApplicationManager`. Only support **kill application** and **query application status** now (codes in `YarnApplicationManager` are directly moved from other files).
- Modify am-related configuration and function name, e.g. change `killYarnJobWhenTaskFailover` to `killApplicationWhenTaskFailover`.
- Add Kubernetes application(pod) status check when submit process exit. Because when deploying in cluster mode on k8s and failures happen inside k8s cluster, submit process will exit normally, which will show wrong status of task execution.

Support submit Spark task directly on Kubernetes(on native Kubernetes, not Spark Operator).

- Modify Spark's resource related configuration to accommodate Spark on K8S deploy mode, e.g. change `--driver-cores` to `--conf spark.driver.cores`.
- Modify relative front-end part.
- Generate master URL based on whether user choose cluster namespace.
             image

- Add `--conf spark.kubernetes.driver.label.dolphinscheduler-label=<processInstanceId>_<taskInstanceId>` automatically for killing driver pod.
- Write cluster's config into `<executePath>/config`, export environment variable `KUBECONFIG` in command file.
- Write driver pod's logs into task's log file, then delete driver pod if it runs successfully.

## Verify this pull request
- Manually tested. BTW, for local test, you have to pull https://github.com/apache/dolphinscheduler/pull/13536 together until it is merged.
       image
        image

### CODE REVIEW

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/ResourceManagerType.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    private final int code;
+    private final String descp;
+
+    public int getCode() {
+        return code;
+    }
+
+    public String getDescp() {
+        return descp;
+    }
```
Comment (kezhenxu94) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/ResourceManagerType.java]: I think all these are unnecessary

**Comment by EricGao888** [comment]
@Radeity If you want to manage the life cycle of spark driver pod and worry about losing driver logs, what about querying the logs of driver pod before deleting it and keep it at DS side? The driver log could be kept as DS spark task log.

**Comment by Radeity** [comment]
> @Radeity If you want to manage the life cycle of spark driver pod and worry about losing driver logs, what about querying the logs of driver pod before deleting it and keep it at DS side? The driver log could be kept as DS spark task log.

Hi, @EricGao888, good idea! Can we directly append write log of driver pod into DS's task log file? In this way, maybe we can delete the driver pod if spark task runs successfully.

**Comment by EricGao888** [comment]
> > @Radeity If you want to manage the life cycle of spark driver pod and worry about losing driver logs, what about querying the logs of driver pod before deleting it and keep it at DS side? The driver log could be kept as DS spark task log.
> 
> Hi, @EricGao888, good idea! Can we directly append write log of driver pod into DS's task log file? In this way, maybe we can delete the driver pod if spark task runs successfully.

Sounds good to me.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManagerContext.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.apache.dolphinscheduler.plugin.task.api.am;
+
+import org.apache.dolphinscheduler.plugin.task.api.K8sTaskExecutionContext;
+
+import lombok.Data;
+
+@Data
+public class KubernetesApplicationManagerContext implements ApplicationManagerContext {
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManagerContext.java:27]: Make this class immutable by adding `final` to the fields and add `@RequiredArgsConstructor` to replace the constructor?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/K8sTaskExecutionContext.java [hunk 3]
```diff
@@ -27,18 +27,29 @@ public class K8sTaskExecutionContext implements Serializable {
...
         return configYaml;
     }
 
+    public void setNamespace(String namespace) {
+        this.namespace = namespace;
+    }
+
+    public String getNamespace() {
+        return namespace;
+    }
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/K8sTaskExecutionContext.java]: I think this class can be immutable, can you revise this?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+     *
+     * @param kubernetesApplicationManagerContext
+     * @return
+     */
+    private KubernetesClient getClient(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) {
+        K8sTaskExecutionContext k8sTaskExecutionContext =
+                kubernetesApplicationManagerContext.getK8sTaskExecutionContext();
+        Config config = Config.fromKubeconfig(k8sTaskExecutionContext.getConfigYaml());
+        return cacheClientMap.computeIfAbsent(kubernetesApplicationManagerContext.getLabelValue(),
+                key -> new DefaultKubernetesClient(config));
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]:
```suggestion
        return cacheClientMap.computeIfAbsent(kubernetesApplicationManagerContext.getLabelValue(),
                key -> new DefaultKubernetesClient(Config.fromKubeconfig(k8sTaskExecutionContext.getConfigYaml())));
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    public void removeCache(String cacheKey) {
+        KubernetesClient client = cacheClientMap.get(cacheKey);
+        if (Objects.nonNull(client)) {
+            client.close();
+        }
+        cacheClientMap.remove(cacheKey);
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]:
```suggestion
        try (KubernetesClient ignored = cacheClientMap.remove(cacheKey)) {
        }
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * collect pod's log
+     *
+     * @param kubernetesApplicationManagerContext
+     * @return
+     */
+    public String collectPodLog(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) {
+        try (KubernetesClient client = getClient(kubernetesApplicationManagerContext)) {
```
Comment (kezhenxu94) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]: Here the `client` is closed automatically, but in `killApplication` (line 192), the `client` will be closed again, (`killApplication` --> `removeCache`)

**Review by kezhenxu94 — COMMENTED** [review]
Thank you @Radeity for opening this PR, I left some comments.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java (hunk 6 above)
Comment (Radeity) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java]: It should be closed in `killApplication`, cuz when master do failover, `collectPodLog` will not be called, i forgot to remove it, thx!

**Comment by Radeity** [comment]
Hi, @kezhenxu94, thanks again for your review, i've already made some changes according to your suggestions, PTAL.

**Comment by rickchengx** [comment]
Hi, @Radeity  Good job! May I ask when the user selects client mode or local mode, will the task be submitted to k8s or executed locally as expected?

**Comment by Radeity** [comment]
> Hi, @Radeity Good job! May I ask when the user selects client mode or local mode, will the task be submitted to k8s or executed locally as expected?


Hi, @rickchengx, thanks for your interest! There are mainly two situations: 
- Select **local** mode: Task runs locally, regardless of namespace selection.
- Select **cluster / client** mode: If user selects namespace either, task will be submitted to **K8S**, instead, to **Yarn**, e.g. when user both selects client mode and namespace, driver will launch inside a pod or on a physical host, executors will launch inside a pod.

**Review by kezhenxu94 — APPROVED** [review]
@EricGao888 wanna recheck?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterConfig.java [hunk 7]
```diff
@@ -88,7 +88,7 @@ public class MasterConfig implements Validator {
     private double maxCpuLoadAvg = -1;
     private double reservedMemory = 0.3;
     private Duration failoverInterval = Duration.ofMinutes(10);
-    private boolean killYarnJobWhenTaskFailover = true;
+    private boolean killApplicationWhenTaskFailover = true;
```
Comment (EricGao888) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterConfig.java:91]: Just curious, why do we set the default value of `killApplicationWhenTaskFailover ` to `true`? If we set it to `false`, will DS continue to track the running application started before and keep idempotence when failover done?

**Review by EricGao888 — APPROVED** [review]
LGTM

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/BaseTaskProcessor.java [hunk 8]
```diff
@@ -635,18 +633,39 @@ public Map<String, String> getResourceFullNames(TaskInstance taskInstance) {
...
+                    namespace = sparkParameters.getNamespace();
+                }
+                break;
+            default:
+                break;
+        }
+
+        if (StringUtils.isNotEmpty(namespace)) {
+            String clusterName = JSONUtils.toMap(namespace).get(CLUSTER);
+            String configYaml = processService.findConfigYamlByName(clusterName);
```
Comment (EricGao888) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/BaseTaskProcessor.java:663]: https://github.com/apache/dolphinscheduler/blob/eaf2541018a1818ad5a0f1cbefe98b095bcc9b8b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java#L2661-L2671
Not sure why method `findConfigYamlByName` was defined in interface `ProcessService`. But this is not related to this PR, we could improve it later.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterConfig.java (hunk 7 above)
Comment (Radeity) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterConfig.java:91]: Get your point, now, DS will not keep tracking running application, and directly set status of taskInstance to  `NEED_FAULT_TOLERANCE`, so the default value sets `true`. Maybe the failover logic can be optimized.

**Comment by EricGao888** [comment]
Will merge this PR if no more comments.

**Pull request merged**